src/request.rs

```diff
@@ -293,11 +294,12 @@ impl<State> Request<State> {
     /// #
     /// # Ok(()) })}
     /// ```
-    pub fn param(&self, key: &str) -> crate::Result<&str> {
+    pub fn param(&self, key: &str) -> crate::Result<String> {
         self.route_params
             .iter()
             .rev()
             .find_map(|params| params.find(key))
+            .map(|param| percent_decode_str(param).decode_utf8_lossy().into())
```

Comment:
If a malformed URL is passed, perhaps we should return a 400 error rather than apply lossy encoding? I would assume that's a case of an invalid URL being passed?

Or do we at this point already know this is valid ASCII so we can continue without a problem?

Reply:
Good point, I'll think a bit more about error handling here and maybe document what we expect to get as input/output.

If we want to return an error here I'm not sure if we should return a 400, or maybe a more specific error that a user can handle?

Reply:
> If we want to return an error here I'm not sure if we should return a 400, or maybe a more specific error that a user can handle?

By user, you mean a user of Tide or a user making a request? In this case I think it's fair to say that the URL sent in the request is invalid, and the server in fact cannot parse it. I'm not sure how else we could handle this on the Tide side.

Reply:
I meant a Tide user. Maybe someone wants some more flexibility around url-parsing-error-handling than just shortcutting to a 400-response. But after looking at the error type we use I don't think we're really set up for that? I think our error type is designed more for reporting errors  to end-users than API-users. So I think that will be a discussion for another time.

I'll fix this to return a 400.

Thanks for the feedback!